### PR TITLE
Modernize components

### PR DESCRIPTION
**Pull request by nhunzaker**
Modernizes a handful of components:

1. Replaces createClass with React.Component
2. Replaces prop types with Flow

### CODE REVIEW

File: src/components/datalist.js [hunk 1]
```diff
@@ -1,32 +1,28 @@
...
+  getOption(record: Record) {
     return <option key={record.id}>{record.caption}</option>
-  },
+  }
 
   render() {
     let { items, id } = this.props
 
-    return <datalist id={id}>{items.map(this.getOption)}</datalist>
+    return <datalist id={id}>{items.map(this.getOption, this)}</datalist>
```
Comment (dbernheisel) [inline src/components/datalist.js]: what's `this` doing here?

File: src/components/figure.js [hunk 2]
```diff
@@ -1,43 +1,40 @@
...
         <span className="ars-fig-caption">{record.name}</span>
         <Ink opacity={0.4} />
       </button>
     )
-  },
+  }
 
-  _onClick(e) {
-    e.preventDefault()
+  _onClick(event: MouseEvent) {
```
Comment (dbernheisel) [inline src/components/figure.js]: any chance this could also be a keyboard event?

File: src/components/datalist.js (hunk 1 above)
Comment (nhunzaker) [reply src/components/datalist.js]: It sets the scope of the `map` callback, so it's sort of like `items.map(this.getOption.bind(this)`, but it doesn't create a copy of the function.

File: src/components/figure.js (hunk 2 above)
Comment (nhunzaker) [reply src/components/figure.js]: Good call. Also this should use the synthetic event system types

Comment (nhunzaker) [reply src/components/figure.js]: I'll do the synthetic typing upstream. For now, I'll generalize this to `Event` since it only needs that interface.

**Pull request merged**